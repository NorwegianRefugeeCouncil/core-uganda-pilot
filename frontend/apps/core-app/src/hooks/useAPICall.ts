import * as React from 'react';

type Parameter<T> = T extends (...args: infer T) => any ? T : never;
type AsyncReturnType<T extends (...args: any) => any> = T extends (
  ...args: any
) => Promise<infer U>
  ? U
  : T extends (...args: any) => infer U
  ? U
  : any;

type State<T> = {
  loading: boolean;
  error: string | null;
  data: T | null;
};

type Action<T> =
  | {
      type: 'INIT';
    }
  | {
      type: 'SUCCESS';
      payload: T;
    }
  | {
      type: 'ERROR';
      payload: string;
    };

const initialState = {
  loading: false,
  error: null,
  data: null,
};

const createDataFetchReducer =
  <T>() =>
  (state: State<T>, action: Action<T>): State<T> => {
    switch (action.type) {
      case 'INIT':
        return {
          ...state,
          loading: true,
        };
      case 'SUCCESS':
        return {
          ...state,
          loading: false,
          error: null,
          data: action.payload,
        };
      case 'ERROR':
        return {
          ...state,
          loading: false,
          error: action.payload,
          data: null,
        };
      default:
        throw new Error();
    }
  };

export const useAPICall = <F extends (...arg: Parameter<F>) => Promise<any>>(
  apiFunction: F,
  args: Parameter<F>,
  rerunOnArgChange: boolean,
): [() => Promise<void>, State<AsyncReturnType<F>>] => {
  const reducer = createDataFetchReducer<AsyncReturnType<F>>();
  const [state, dispatch] = React.useReducer(reducer, initialState);

  const run = React.useCallback(async () => {
    dispatch({ type: 'INIT' });
    try {
      const data = await apiFunction(...args);
      dispatch({ type: 'SUCCESS', payload: data });
    } catch (error: any) {
      dispatch({ type: 'ERROR', payload: error?.message ?? 'Unknown error' });
    }
  }, [JSON.stringify(args)]);

  React.useEffect(() => {
    if (rerunOnArgChange) {
      run();
    }
  }, [JSON.stringify(args)]);

  return [run, state];
};
