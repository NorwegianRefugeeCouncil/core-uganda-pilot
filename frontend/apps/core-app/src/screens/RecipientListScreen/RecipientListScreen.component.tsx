import * as React from 'react';

import { RecordTableContext } from '../../components/RecordTable/RecordTableContext';
import { data } from '../../components/RecordTable/data.tmp';
import RecordTable from '../../components/RecordTable';
import { GlobalTableFilter } from '../../components/RecordTable/GlobalTableFilter';

import * as Styles from './RecipientListScreen.styles';

type Props = {
  onItemClick: (id: string) => void;
};

export const RecipientListScreenComponent: React.FC<Props> = ({
  onItemClick,
}) => {
  const tableContext = React.useContext(RecordTableContext);

  return (
    <Styles.Container>
      {tableContext?.tableInstance && (
        <GlobalTableFilter
          table={tableContext.tableInstance}
          globalFilter={tableContext.tableInstance.globalFilter}
          setGlobalFilter={tableContext.tableInstance.setGlobalFilter}
        />
      )}
      <RecordTable data={data} onItemClick={onItemClick} />
    </Styles.Container>
  );
};
