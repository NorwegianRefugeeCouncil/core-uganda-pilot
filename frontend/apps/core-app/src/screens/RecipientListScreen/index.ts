export { RecipientListScreenContainer as RecipientListScreen } from './RecipientListScreen.container';
