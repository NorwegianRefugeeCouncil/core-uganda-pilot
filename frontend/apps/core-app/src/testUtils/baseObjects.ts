import { FormDefinition, FormType, Record } from 'core-api-client';

export const baseForm: FormDefinition = {
  id: 'form1',
  code: '',
  name: 'form 1',
  databaseId: 'databaseId1',
  folderId: 'folderId1',
  formType: FormType.DefaultFormType,
  fields: [],
};

export const baseTextField = {
  id: 'field1',
  name: 'field 1',
  description: 'description 1',
  code: '',
  required: false,
  key: false,
  fieldType: {
    text: {},
  },
};

export const baseMultilineTextField = {
  id: 'field2',
  name: 'field 2',
  description: 'description 2',
  code: '',
  required: true,
  key: true,
  fieldType: {
    multilineText: {},
  },
};

export const baseRecord1: Record = {
  id: 'record1',
  values: [
    {
      value: 'text',
      fieldId: 'field1',
    },
    {
      value: 'multi line text',
      fieldId: 'field2',
    },
  ],
  formId: 'form1',
  databaseId: 'databaseId1',
  ownerId: undefined,
};
