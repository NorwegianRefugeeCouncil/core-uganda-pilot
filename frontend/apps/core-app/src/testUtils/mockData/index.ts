export { makeRecord } from './Record';
export { makeForm } from './Form';
export { makeField } from './FormField';
