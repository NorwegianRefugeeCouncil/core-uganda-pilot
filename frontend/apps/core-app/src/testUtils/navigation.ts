export const mockNavigationProp = {
  dispatch: jest.fn(),
  navigate: jest.fn(),
  reset: jest.fn(),
  goBack: jest.fn(),
  isFocused: jest.fn(),
  canGoBack: jest.fn(),
  getParent: jest.fn(),
  getState: jest.fn(),
  setParams: jest.fn(),
  setOptions: jest.fn(),
  addListener: jest.fn(),
  removeListener: jest.fn(),
  replace: jest.fn(),
  push: jest.fn(),
  pop: jest.fn(),
  popToTop: jest.fn(),
};
