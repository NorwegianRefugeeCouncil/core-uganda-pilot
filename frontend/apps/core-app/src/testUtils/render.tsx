import * as React from 'react';
import { NavigationContainer } from '@react-navigation/native';
import {
  render as r,
  RenderAPI,
  RenderOptions,
} from '@testing-library/react-native';
import { render as rWeb } from '@testing-library/react';
import { NativeBaseProvider } from 'native-base';
import { theme } from 'core-design-system';

const inset = {
  frame: { x: 0, y: 0, width: 0, height: 0 },
  insets: { top: 0, left: 0, right: 0, bottom: 0 },
};

export const render = (
  component: React.ReactElement<any>,
  options?: RenderOptions | undefined,
): RenderAPI =>
  r(
    <NativeBaseProvider theme={theme} initialWindowMetrics={inset}>
      <NavigationContainer>{component}</NavigationContainer>
    </NativeBaseProvider>,
    options,
  );

export const renderWeb = (
  component: React.ReactElement<any>,
  options?: RenderOptions | undefined,
) =>
  rWeb(
    <NativeBaseProvider theme={theme} initialWindowMetrics={inset}>
      <NavigationContainer>{component}</NavigationContainer>
    </NativeBaseProvider>,
    options,
  );
