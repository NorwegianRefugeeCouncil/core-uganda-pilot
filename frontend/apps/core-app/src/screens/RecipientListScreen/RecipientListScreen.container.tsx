import * as React from 'react';
import { StackScreenProps } from '@react-navigation/stack';

import { RootNavigatorParamList } from '../../navigation/root';
import { useRecipientForms } from '../../contexts/RecipientForms';

import { RecipientListScreenComponent } from './RecipientListScreen.component';

type Props = StackScreenProps<RootNavigatorParamList, 'recipientsList'>;

export const RecipientListScreenContainer: React.FC<Props> = () => {
  const recipientForms = useRecipientForms();

  return (
    <RecipientListScreenComponent
      forms={recipientForms}
      isLoading={!recipientForms || !recipientForms.length}
    />
  );
};
