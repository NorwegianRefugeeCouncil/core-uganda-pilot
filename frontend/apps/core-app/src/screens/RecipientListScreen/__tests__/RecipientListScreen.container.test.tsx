import React from 'react';

import { render } from '../../../testUtils/render';
import { RecipientListScreenContainer } from '../RecipientListScreen.container';
import * as hooks from '../../../hooks/useAPICall';

const mockNavigate = jest.fn();
// const useRecipientFormsMock = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockNavigate,
    }),
  };
});

// jest.mock('../../../contexts/RecipientForms', () => {
//   return {
//     useRecipientForms: useRecipientFormsMock,
//   };
// });

describe('RecipientListScreenContainer', () => {
  const useAPICallSpy = jest.spyOn(hooks, 'useAPICall');

  afterEach(() => {
    useAPICallSpy.mockReset();
  });

  it('should match the snapshot', () => {
    const { toJSON } = render(<RecipientListScreenContainer />);
    expect(toJSON()).toMatchSnapshot();
  });

  // it('should call useRecipientForms', () => {
  //   render(<RecipientListScreenContainer />);
  //   expect(useRecipientFormsMock).toHaveBeenCalledTimes(1);
  // });
});
