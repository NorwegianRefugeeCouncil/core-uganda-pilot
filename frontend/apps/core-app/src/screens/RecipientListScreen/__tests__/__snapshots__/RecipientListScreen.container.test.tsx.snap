// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipientListScreenContainer should match the snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    _text={
      Object {
        "color": "darkText",
      }
    }
    contentContainerStyle={
      Array [
        Object {},
        Object {
          "dataSet": Object {},
        },
      ]
    }
    dataSet={Object {}}
    style={
      Object {
        "backgroundColor": "#fff",
      }
    }
  >
    <View>
      <View
        dataSet={Object {}}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F6F6F7",
            "marginBottom": 16,
            "marginTop": 16,
            "width": "100%",
          }
        }
      >
        <View
          dataSet={Object {}}
          style={
            Object {
              "maxWidth": 1180,
              "width": "100%",
            }
          }
        >
          <View
            dataSet={Object {}}
            style={
              Object {
                "marginBottom": 42,
                "marginRight": "auto",
                "marginTop": 26,
                "maxWidth": 580,
                "width": "100%",
              }
            }
          >
            <View
              dataSet={Object {}}
              style={
                Object {
                  "width": "100%",
                }
              }
            >
              <View
                dataSet={Object {}}
                feedbackId="field-react-aria-1-feedback"
                hasFeedbackText={false}
                hasHelpText={false}
                helpTextId="field-react-aria-1-helptext"
                isDisabled={false}
                isInvalid={false}
                isReadOnly={false}
                isRequired={false}
                labelId="field-react-aria-1-label"
                nativeID="field-react-aria-1-label"
                setHasFeedbackText={[Function]}
                setHasHelpText={[Function]}
                style={
                  Object {
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "marginBottom": 2,
                    "marginRight": 12,
                  }
                }
              >
                <Text
                  dataSet={Object {}}
                  level="1"
                  style={
                    Object {
                      "backgroundColor": undefined,
                      "color": "#666666",
                      "fontFamily": "Roboto_500Medium",
                      "fontSize": 18,
                      "fontStyle": undefined,
                      "fontWeight": undefined,
                      "letterSpacing": 0,
                      "lineHeight": 26,
                      "textAlign": "auto",
                      "textDecorationLine": undefined,
                    }
                  }
                >
                  Beneficiary Name
                </Text>
              </View>
              <TextInput
                _text={
                  Object {
                    "color": "darkText",
                  }
                }
                accessible={true}
                allowFontScaling={true}
                dataSet={Object {}}
                disabled={false}
                editable={true}
                isRequired={false}
                nativeID="field-react-aria-1"
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                onKeyPress={[Function]}
                placeholder="Search"
                placeholderTextColor="#a3a3a3"
                readOnly={false}
                rejectResponderTermination={true}
                required={false}
                secureTextEntry={false}
                selectionColor="#1f2937"
                style={
                  Array [
                    Object {
                      "backgroundColor": "#fff",
                      "borderColor": "#666666",
                      "borderRadius": 3,
                      "borderStyle": "solid",
                      "borderWidth": 1,
                      "color": "#666666",
                      "fontFamily": "Roboto_400Regular",
                      "fontSize": 18,
                      "fontStyle": undefined,
                      "fontWeight": undefined,
                      "lineHeight": 26,
                      "overflow": "hidden",
                      "paddingBottom": 12,
                      "paddingLeft": 12,
                      "paddingRight": 12,
                      "paddingTop": 12,
                      "width": undefined,
                    },
                    Object {},
                  ]
                }
                testID="recipient-list-table-filter"
                underlineColorAndroid="transparent"
                value=""
              />
            </View>
          </View>
        </View>
      </View>
      <View
        dataSet={Object {}}
        style={
          Object {
            "marginLeft": "auto",
            "marginRight": "auto",
            "maxWidth": 1180,
            "width": "100%",
          }
        }
      >
        <View
          dataSet={Object {}}
          fadeDuration={0.1}
          speed={1}
          style={
            Object {
              "backgroundColor": "transparent",
              "height": 160,
              "overflow": "hidden",
              "paddingBottom": 16,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 16,
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "#e5e5e5",
                "height": "100%",
                "opacity": 0,
                "width": "100%",
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaProvider>
`;
